pallets/msa/src/lib.rs

```diff
@@ -883,6 +976,90 @@ impl<T: Config> Pallet<T> {
 		}
 		Ok(Some(schemas.0.into()))
 	}
+
+	/// Add a signature to the PayloadSignatureRegistry based on a virtual "bucket" grouping.
+	/// First reset the virtual bucket if necessary, then add the new entry.
+	pub fn register_signature(
+		signature: &MultiSignature,
+		current_block: T::BlockNumber,
+		mortality_block: T::BlockNumber,
+	) -> DispatchResult {
+		if Self::mortality_block_limit(current_block) < mortality_block {
+			Err(Error::<T>::MortalityTooHigh.into())
+		} else {
+			let bucket_num = Self::bucket_for(mortality_block.into());
+			<PayloadSignatureRegistry<T>>::try_mutate(
+				bucket_num,
+				signature,
+				|maybe_mortality_block| -> DispatchResult {
+					ensure!(maybe_mortality_block.is_none(), Error::<T>::BucketKeyExists);
+
+					Self::reset_virtual_bucket_if_needed(mortality_block, bucket_num)?;
+					*maybe_mortality_block = Some(mortality_block);
+					Ok(())
+				},
+			)
+		}
+	}
+
+	// Check if enough blocks have passed to reset bucket mortality storage.
+	// If so:
+	//     1. delete all the stored bucket/signature alues with key1 = bucket num
+	//     2. set the bucket's mortality block to the new value
+	// If not, don't do anything.
+	fn reset_virtual_bucket_if_needed(
+		mortality_block: T::BlockNumber,
+		bucket_num: T::BlockNumber,
+	) -> DispatchResult {
+		<MortalityBlockOf<T>>::try_mutate(bucket_num, |maybe_mortality_block| -> DispatchResult {
+			let bucket_mortality_block = Self::bucket_mortality_block(mortality_block);
+
+			match maybe_mortality_block {
+				None => Err(Error::<T>::NoSuchBucket.into()),
+				Some(mortality_block_for_bucket) => {
+					if mortality_block > *mortality_block_for_bucket {
+						Self::delete_signatures_for_bucket(bucket_num)?;
+						*maybe_mortality_block = Some(bucket_mortality_block);
+					} else if *mortality_block_for_bucket == T::BlockNumber::default() {
+						*maybe_mortality_block = Some(bucket_mortality_block);
+					}
+					Ok(())
+				},
+			}
+		})
+	}
+
+	// delete signatures with the given bucket number since they have existed past the mortality
+	// block limit.
+	fn delete_signatures_for_bucket(bucket_num: T::BlockNumber) -> Result<(), DispatchError> {
+		// use drain_prefix
+		// https://paritytech.github.io/substrate/master/frame_support/pallet_prelude/struct.StorageDoubleMap.html#method.drain_prefix
+		// we dont' need the value; it's just to finalize the drain.
+		<PayloadSignatureRegistry<T>>::drain_prefix(bucket_num).count();
+		Ok(())
+	}
+
+	// Virtual bucket mortalities are increasing multiples of MortalityBucketSize.
+	fn bucket_mortality_block(for_block: T::BlockNumber) -> T::BlockNumber {
+		let bucket_size: T::BlockNumber = T::MortalityBucketSize::get().into();
+		(for_block / bucket_size + T::BlockNumber::from(1u32)) * bucket_size
```

Comment:
```suggestion
		(for_block / bucket_size + T::BlockNumber::one()) * bucket_size
```

Reply:
I tried that already, unfortunately:
```
 996 |         (for_block / bucket_size + T::BlockNumber::one() * bucket_size
      |                                                    ^^^ function or associated item not found in `<T as frame_system::Config>::BlockNumber`
```

Reply:
@shannonwells I wonder if instead you can operate with u32 all the way to the end and then do an into(). Instead of doing the math on the `BlockNumber` type. Just a thought.

Reply:
Not without other type gymnastics, no.  And we also actually want some of these to be BlockNumbers.  I was only willing to spend so much time arguing with rust.

Reply:
Good point, if we mutate Blocknumber to do an add , saturating_add, not sure if its a type trait

Reply:
read the while code the whole output can be T::BlockNumber::from({/// some value u32}) or an into might work too

Reply:
it's already using `.one()`, I just needed to use the `One` trait at the top.